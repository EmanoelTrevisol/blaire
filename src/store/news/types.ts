import { INews } from '../../models/News';

export enum ActionTypes {
  GET_NEWS = 'GET_NEWS',
  SET_NEWS_LIST = 'SET_NEWS_LIST',
}

export interface IGetNewsActions {
  type: ActionTypes.GET_NEWS;
}

export interface ISetNewsListActions {
  type: ActionTypes.SET_NEWS_LIST;
  payload: {
    news: INews[];
  };
}

export type NewsActionTypes = IGetNewsActions | ISetNewsListActions;
