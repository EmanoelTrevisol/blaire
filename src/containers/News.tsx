import React from 'react';
import { View, Text } from 'react-native';

const News = () => {
  return (
    <View>
      <Text>Novidades</Text>
    </View>
  );
};

export default News;
