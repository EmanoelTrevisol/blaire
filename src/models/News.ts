import moment from 'moment';

export interface INewsUser {
  name: string;
  profile_picture: string;
}

export interface INewsMessage {
  content: string;
  created_at: string;
}

export interface INewsApiResponse {
  user: INewsUser;
  message: INewsMessage;
}

export interface INews {
  message: string;
  createdAt: string;
  username: string;
  profilePicture: string;
  key: string;
}

export default class News implements INews {
  message: string;
  username: string;
  profilePicture: string;
  createdAt: string;
  key: string;

  constructor({ user, message }: INewsApiResponse) {
    this.message = message.content;
    this.createdAt = moment(message.created_at).format('lll');
    this.key = message.created_at;
    this.username = user.name;
    this.profilePicture =
      'https://pbs.twimg.com/profile_images/1323222831065374722/dldoUekJ_400x400.jpg';

    // this.profilePicture = user.profile_picture;
    // the profile_picture in the response does not exist anymore so I changed it to have a real image
  }
}
