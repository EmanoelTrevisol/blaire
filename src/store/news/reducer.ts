import initialState from './state';
import { ActionTypes, NewsActionTypes } from './types';

export default function (state = initialState, action: NewsActionTypes) {
  switch (action.type) {
    case ActionTypes.SET_NEWS_LIST:
      return {
        ...state,
        list: action.payload.news,
      };
    default:
      return state;
  }
}
