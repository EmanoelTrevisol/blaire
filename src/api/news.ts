import axios from 'axios';
import { apiPaths } from '../config';
import { INewsApiResponse } from '../models/News';

export interface IGetNewsResponse {
  news: INewsApiResponse[];
}

export function getNews(): Promise<IGetNewsResponse> {
  return axios.get(apiPaths.news);
}
