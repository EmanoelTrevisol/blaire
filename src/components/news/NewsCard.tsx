import React from 'react';
import { View, StyleSheet, Text, Image } from 'react-native';

import { INews } from '../../models/News';

const NewsCard = (props: INews) => {
  const { username, createdAt, profilePicture, message } = props;

  return (
    <View style={stl.card}>
      <View style={stl.header}>
        <Image style={stl.profilePicture} source={{ uri: profilePicture }} />
        <Text style={stl.username}>{username}</Text>
        <Text style={stl.createdAt}>{createdAt}</Text>
      </View>
      <View style={stl.body}>
        <Text style={stl.bodyText}>{message}</Text>
      </View>
    </View>
  );
};

const stl = StyleSheet.create({
  card: {
    backgroundColor: '#fff',
    paddingVertical: 10,
    marginHorizontal: 15,
    marginBottom: 15,
    borderRadius: 20,
    elevation: 10,
    shadowColor: '#000',
    shadowRadius: 3,
    shadowOpacity: 0.2,
    shadowOffset: {
      width: 0,
      height: 4,
    },
  },
  header: {
    padding: 20,
    justifyContent: 'center',
    alignItems: 'center',
    borderBottomColor: '#DADADA',
    borderBottomWidth: 1,
  },
  profilePicture: {
    height: 80,
    width: 80,
    borderRadius: 40,
    marginBottom: 5,
  },
  username: {
    fontSize: 24,
    fontWeight: '600',
  },
  createdAt: {
    fontSize: 12,
    color: '#444',
  },
  body: {
    padding: 20,
    // paddingHorizontal: 20,
    // paddingVertical: 16,
  },
  bodyText: {
    fontSize: 16,
  },
});

export default NewsCard;
