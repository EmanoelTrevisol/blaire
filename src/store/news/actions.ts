import { ActionTypes } from './types';
import { getNews } from '../../api/news';
import { setNewsLoading } from '../loader/actions';
import News, { INews } from '../../models/News';
import { sortNewsByDate } from '../../utils/sort';

export function getLatestNews() {
  return async (dispatch) => {
    try {
      dispatch(setNewsLoading(true));

      const res = await getNews();

      const news = sortNewsByDate(res.news).map((n) => new News({ ...n }));

      dispatch(setNewsList(news));
    } catch (error) {
      // TODO: treat error
      console.log(error);
    } finally {
      dispatch(setNewsLoading(false));
    }
  };
}

export function setNewsList(list: INews[]) {
  return {
    type: ActionTypes.SET_NEWS_LIST,
    payload: {
      news: list,
    },
  };
}
