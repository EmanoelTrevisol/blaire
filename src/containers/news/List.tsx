import React, { useCallback, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import {
  View,
  SafeAreaView,
  FlatList,
  StyleSheet,
  Platform,
  RefreshControl,
  Image,
} from 'react-native';
import { ThunkAction } from 'redux-thunk';
import { treatError } from '../../errors/handler';
import News from '../../models/News';
import NewsCard from '../../components/news/NewsCard';
import { getLatestNews } from '../../store/news/actions';
import Loader from '../../components/Loader';

export interface IProps {
  list: News[];
  dispatch: ThunkAction;
  loading: boolean;
}

const ListHeaderComponent = () => (
  <View>
    <Image source={require('../../assets/images/grupo-boticario-logo.png')} />
  </View>
);

const List = (props: IProps) => {
  const [isRefreshing, setIsRefreshing] = useState<boolean>(false);

  const { list, dispatch, loading } = props;

  const loadLatestNews = useCallback(
    async (isRefresh: boolean = false) => {
      try {
        if (isRefresh) {
          setIsRefreshing(true);
        }
        await dispatch(getLatestNews());
      } catch (error) {
        return treatError(
          error,
          'Tivemos um problema ao carregar as novidades. Por favor, tente novamente',
        );
      } finally {
        if (isRefresh) {
          setIsRefreshing(false);
        }
      }
    },
    [dispatch],
  );

  useEffect(() => {
    loadLatestNews(false);
  }, [loadLatestNews]);

  if (loading && !isRefreshing) {
    return <Loader />;
  }

  return (
    <SafeAreaView>
      <View style={stl.container}>
        <FlatList
          data={list}
          refreshing={isRefreshing}
          onRefresh={() => loadLatestNews(true)}
          keyExtractor={(item) => item.key}
          renderItem={({ item }) => <NewsCard {...item} />}
          refreshControl={
            <RefreshControl
              refreshing={isRefreshing}
              onRefresh={() => loadLatestNews(true)}
              tintColor="#000"
            />
          }
          ListHeaderComponent={ListHeaderComponent}
          ListHeaderComponentStyle={stl.listHeaderStyle}
        />
      </View>
    </SafeAreaView>
  );
};

const stl = StyleSheet.create({
  container: {
    ...Platform.select({
      android: {
        marginTop: 15,
      },
    }),
  },
  listHeaderStyle: {
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 25,
  },
});

export default connect((state) => ({
  loading: state.loader.news,
  list: state.news.list,
}))(List);
