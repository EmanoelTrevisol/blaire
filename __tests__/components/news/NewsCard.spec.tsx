import React from 'react';
import { render } from '@testing-library/react-native';
import { Image } from 'react-native';

import NewsCard from '@components/news/NewsCard';
import { news } from '@models/fakeData/news';
import { INews } from '@models/News';

let props: INews;

describe('NewsCard.tsx', () => {
  const renderComp = (cpProps: INews) => {
    return render(<NewsCard {...cpProps} />);
  };

  beforeEach(() => {
    props = { ...news[0] };
  });

  test('matches snapshot', () => {
    const { toJSON } = renderComp(props);

    expect(toJSON()).toMatchSnapshot();
  });

  test('has props in the document', () => {
    const { getByText, UNSAFE_getByType } = renderComp(props);

    const profilePicture = UNSAFE_getByType(Image);

    expect(profilePicture).toBeDefined();
    expect(profilePicture.props.source).toEqual({
      uri: props.profilePicture,
    });
    expect(getByText(props.message)).toBeDefined();
    expect(getByText(props.createdAt)).toBeDefined();
    expect(getByText(props.username)).toBeDefined();
    expect(getByText(props.username)).toBeDefined();
  });
});
