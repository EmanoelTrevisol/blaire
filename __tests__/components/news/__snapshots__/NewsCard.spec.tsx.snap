// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsCard.tsx matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 20,
      "elevation": 10,
      "marginBottom": 15,
      "marginHorizontal": 15,
      "paddingVertical": 10,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 4,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#DADADA",
        "borderBottomWidth": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://pbs.twimg.com/profile_images/1323222831065374722/dldoUekJ_400x400.jpg",
        }
      }
      style={
        Object {
          "borderRadius": 40,
          "height": 80,
          "marginBottom": 5,
          "width": 80,
        }
      }
    />
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "600",
        }
      }
    >
      O Boticário
    </Text>
    <Text
      style={
        Object {
          "color": "#444",
          "fontSize": 12,
        }
      }
    >
      Feb 2, 2020 1:10 PM
    </Text>
  </View>
  <View
    style={
      Object {
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
    >
      Além disso, nossos funcionários e familiares receberão kits de proteção. Afinal, o cuidado começa aqui dentro, né?
    </Text>
  </View>
</View>
`;
